Create SimpleChannelEventHandler convenience class

Bennett-Lynch opened the pull request:
Motivation:

Currently, the vast majority of userEventTriggered() implementations
require the user to supply the boilerplate behavior of performing an
instanceof check, handling if appropriate, and calling
fireUserEventTriggered() otherwise.

We can simplify this very common use case by creating a class that only
matches user events of a given type, similar to the existing
SimpleChannelInboundHandler class.

Modifications:

Create a new SimpleChannelEventHandler class
Create accompanying SimpleChannelEventHandlerTest

Result:

Users will be able to handle most events in a less verbose manner.

Bennett-Lynch commented:
Note: Open to better naming suggestions for "SimpleChannelEventHandler" and "eventReceived".

normanmaurer commented on transport/src/main/java/io/netty/channel/SimpleChannelEventHandler.java:
```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2013 The Netty Project
```
2018

Bennett-Lynch replied:
Fixing.

normanmaurer commented on transport/src/main/java/io/netty/channel/SimpleChannelEventHandler.java:
```diff
@@ -0,0 +1,85 @@
+     * Returns {@code true} if the given user event should be handled. If {@code false} it will be passed to the next
+     * {@link ChannelInboundHandler} in the {@link ChannelPipeline}.
+     */
+    public boolean acceptEvent(Object evt) throws Exception {
```
protected ?

Bennett-Lynch replied:
Protected makes more sense to me.  I followed the design of `SimpleChannelInboundHandler`'s `acceptInboundMessage`, which is declared public.

normanmaurer commented on transport/src/main/java/io/netty/channel/SimpleChannelEventHandler.java:
```diff
@@ -0,0 +1,85 @@
+        if (acceptEvent(evt)) {
+            @SuppressWarnings("unchecked")
+            I ievt = (I) evt;
+            eventReceived(ctx, ievt);
```
I think we should follow the same pattern as in `SimpleChannelInboundHandler` and allow to automatically call `release` on the event.

Bennett-Lynch replied:
I wasn't sure how often `release` is applicable to events, but I agree it probably makes more sense, and it's more consistent with the "Simple" naming convention.  Changing.

normanmaurer commented on transport/src/main/java/io/netty/channel/SimpleChannelEventHandler.java:
```diff
@@ -0,0 +1,85 @@
+ *     }
+ * </pre>
+ */
+public abstract class SimpleChannelEventHandler<I> extends ChannelInboundHandlerAdapter {
```
`SimpleUserEventChannelHandler` (as user events are inbound all the time)?

Bennett-Lynch replied:
Neither name implies a tendency towards or against "inbound", do they?  But either way, I agree that name sounds better.  Changing.

normanmaurer commented on transport/src/test/java/io/netty/channel/SimpleChannelEventHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2013 The Netty Project
```
2018

Bennett-Lynch replied:
Fixing.

normanmaurer commented on transport/src/test/java/io/netty/channel/SimpleChannelEventHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+        assertTrue(allEventCatcher.caughtEvents.size() == 1);
+    }
+
+    private static class FooEvent {
```
final

normanmaurer commented on transport/src/test/java/io/netty/channel/SimpleChannelEventHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+    private static class FooEvent {
+    }
+
+    private class FooEventCatcher extends SimpleChannelEventHandler<FooEvent> {
```
final

normanmaurer commented on transport/src/test/java/io/netty/channel/SimpleChannelEventHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+
+    private class AllEventCatcher extends ChannelInboundHandlerAdapter {
```
final

normanmaurer commented on transport/src/test/java/io/netty/channel/SimpleChannelEventHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+            caughtEvents.add(evt);
+        }
+    }
+}
```
add line break

normanmaurer commented on transport/src/test/java/io/netty/channel/SimpleChannelEventHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        @Override
+        public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
```
remove `throws...`

normanmaurer commented on transport/src/test/java/io/netty/channel/SimpleChannelEventHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        @Override
+        protected void eventReceived(ChannelHandlerContext ctx, FooEvent evt) throws Exception {
```
remove `throws...`

normanmaurer commented on transport/src/test/java/io/netty/channel/SimpleChannelEventHandlerTest.java:
```diff
@@ -0,0 +1,86 @@
+    public void setUp() throws Exception {
+        fooEventCatcher = new FooEventCatcher();
+        allEventCatcher = new AllEventCatcher();
+        channel = new EmbeddedChannel(fooEventCatcher, allEventCatcher);
```
call channel.finish() after test execution is done and verify return value

Bennett-Lynch commented:
Addressed all feedback.

Question: Should we declare `userEventTriggered` to be `final`?  It makes sense to me, and makes usages of the class less error-prone.  `SimpleChannelInboundHandler` does not do this today though.

normanmaurer commented:
@Bennett-Lynch yeah final sounds good

Bennett-Lynch commented:
Added final and addressed all feedback.

Do we need more people to weigh in with their opinion on the class being worthwhile?

bryce-anderson commented on transport/src/test/java/io/netty/channel/SimpleUserEventChannelHandlerTest.java:
```diff
@@ -0,0 +1,101 @@
+        channel.pipeline().fireUserEventTriggered(barEvent);
+        assertEquals(0, fooEventCatcher.caughtEvents.size());
+        assertEquals(1, allEventCatcher.caughtEvents.size());
+        assertEquals(1, barEvent.refCnt());
```
We need to release this event; It could either be explicit here or as part of the lifecycle of `AllEventHandler`.

Bennett-Lynch replied:
Fixed, thanks.

dschobel commented:
> Do we need more people to weigh in with their opinion on the class being worthwhile?

this looks useful to me, lgtm after the other feedback

Bennett-Lynch commented:
Made test helper classes package private.

Bennett-Lynch commented:
@normanmaurer @ejona86 Any additional feedback?

normanmaurer commented:
@Bennett-Lynch Please squash and update the commit message if needed.... after done I will merge.

Thanks for your patience

Bennett-Lynch commented:
@normanmaurer Done, thank you.
